### PR TITLE
Enable type hints and rich linking in Sphinx docs

### PR DESCRIPTION
**Pull request by itsdawei**
## Description

Enable type hints and rich linking in Sphinx docs so we can use python3 type hints in our code as opposed to adding types in docstrings by hand.
For demonstration, migrate `ribs/visualize.py` to use python type hint.

## TODO

- [x] Upgrade to Sphinx 4.5.0 (see https://github.com/tox-dev/sphinx-autodoc-typehints/issues/205#issuecomment-1008837900)
- [x] Migrate `ribs/visualize.py` to use python type hints.
- [x] Fix compatibility issues caused by the upgrade to 4.5.0

Waiting for `sphinx-toolbox == 3.0.0` compatibility issue with `sphinx-autodoc-typehints == 1.18.2` to be fixed.

## Status

- [x] I have read the guidelines in [CONTRIBUTING.md](https://github.com/icaros-usc/pyribs/blob/master/CONTRIBUTING.md).
- [x] I have formatted my code using `yapf`
- [x] I have tested my code by running `pytest`
- [x] I have linted my code with `pylint`
- [x] I have added a one-line description of my change to the changelog in `HISTORY.md`.
- [ ] This PR is ready to go

### CODE REVIEW

**Review by btjanaka — CHANGES_REQUESTED** [review]
Fix dep problems so that we pass tests

File: ribs/visualize.py [hunk 1]
```diff
@@ -81,21 +85,23 @@ def _set_cbar(t, ax, cbar, cbar_kwargs):
...
+                         cmap: Union[str, list,
+                                     matplotlib.colors.Colormap] = "magma",
+                         square: bool = None,
+                         aspect: Union['auto', 'equal', float] = None,
+                         vmin: float = None,
+                         vmax: float = None,
+                         cbar: Union['auto', None,
+                                     matplotlib.axes.Axes] = "auto",
+                         pcm_kwargs: dict = None,
+                         cbar_kwargs: dict = None) -> GridArchive:
```
Comment (itsdawei) [inline ribs/visualize.py:104]: This line is added to test if return type hints work. It does. 
- [ ] Remember to remove this line before merging.

**Comment by itsdawei** [comment]
This is what the documentation page looks like now for `ribs.visualize.grid_archive_heatmap`
image

**Comment by itsdawei** [comment]
We are closing this pull request for now for the following reasons:
- The typing module is extremely volatile https://docs.python.org/3/library/typing.html -- things are being deprecated and introduced left and right; `typing.List` was even deprecated in Python 3.9
- Type hints may introduce additional unnecessary computation -- the types themselves require computation, and more severely, we may have to make unnecessary imports just to annotate existing code e.g. in `ribs.visualize`. There is no such computation when you put the types in comments.
- Type hints may be unreadable to new Python users -- e.g. `Union` and `Optional`

**Pull request closed without merging**